src/zx_plot.jl

```diff
@@ -0,0 +1,208 @@
+                v = lo.spider_seq[q][frontier_v[q]]
+                nb = neighbors(zxd, v)
+                if length(nb) <= 2
+                    locs[v] = (Float64(frontier_locs[q]), Float64(q))
+                    frontier_locs[q] += 1
+                    frontier_v[q] += 1
+                else
+                    v1 = nb[[qubit_loc(lo, u) != q for u in nb]][1]
+                    if spider_type(zxd, v1) == SpiderType.H
+                        v1 = setdiff(neighbors(zxd, v1), [v])[1]
+                    end
+                    if sum([findfirst(isequal(u), lo.spider_seq[qubit_loc(lo, u)]) != frontier_v[qubit_loc(lo, u)] for u in [v, v1]]) == 0
+                        x = maximum(frontier_locs[min(qubit_loc(lo, v), qubit_loc(lo, v1)):max(qubit_loc(lo, v), qubit_loc(lo, v1))])
+                        for u in [v, v1]
+                            locs[u] = (Float64(x), Float64(qubit_loc(lo, u)))
+                            frontier_v[qubit_loc(lo, u)] += 1
+                        end
+                        for q in min(qubit_loc(lo, v), qubit_loc(lo, v1)):max(qubit_loc(lo, v), qubit_loc(lo, v1))
+                            frontier_locs[q] = x + 1
+                        end
+                    end
+                end
+            end
+        end
+    end
+    for v in vs
+        if !haskey(locs, v)
+            v1, v2 = neighbors(zxd, v)
+            x1, y1 = locs[v1]
+            x2, y2 = locs[v2]
+            locs[v] = ((x1+x2)/2, (y1+y2)/2)
+        end
+    end
+    locs_x = [locs[v][1] for v in vs]
+    locs_y = [locs[v][2] for v in vs]
+    return locs_x, locs_y
+end
+
+function layout2locs(zxd::ZXGraph{T,P}) where {T,P}
+    lo = zxd.layout
+    vs = spiders(zxd)
+    locs = Dict()
+    nqubit = lo.nbits
+    frontier_v = ones(T, nqubit)
+    frontier_locs = ones(nqubit)
+    phase_gadget_loc = 1.0
+
+    for v in vs
+        if qubit_loc(lo, v) != nothing
+            y = qubit_loc(lo, v)
+            x = findfirst(isequal(v), lo.spider_seq[y])
+            locs[v] = (Float64(x), Float64(y))
+        else
+            locs[v] = nothing
+        end
+    end
+    for v in vs
+        if locs[v] == nothing
+            nb = neighbors(zxd, v)
+            if length(nb) == 1
+                u = nb[1]
+                locs[v] = (phase_gadget_loc, Float64(nqubit + 2))
+                locs[u] = (phase_gadget_loc, Float64(nqubit + 1))
+                phase_gadget_loc += 1
+            end
+
+            # v1, v2 = neighbors(zxd, v)
+            # x1, y1 = locs[v1]
+            # x2, y2 = locs[v2]
+            # locs[v] = ((x1+x2)/2, (y1+y2)/2)
+        end
+    end
+    println(locs)
+    locs_x = [locs[v][1] for v in vs]
+    locs_y = [locs[v][2] for v in vs]
+    return locs_x, locs_y
+end
+
+function ZXplot(zxd::ZXDiagram; linetype = "straight")
+    g, edgelabel = ZX2Graph(zxd)
+    nodelabel = ZX2nodelabel(zxd)
+    nodefillc = ZX2nodefillc(zxd)
+    edgelabelc = colorant"black"
+    if zxd.layout.nbits > 0
```

Comment:
try to remove some duplicated code, so it will be easier to maintain later. The following part is a bit duplicated with the one for ZXGraph

Reply:
The functions used for ZXDiagram and ZXGraph are different.